common/beerocks/bcl/include/bcl/network/socket_refactorings.h

```diff
@@ -0,0 +1,795 @@
+    }
+
+    /**
+     * Delete copy constructor to avoid having multiple copies of the file descriptor
+     */
+    FileDescriptorImpl(const FileDescriptorImpl &) = delete;
+
+    /**
+     * Delete assignment operator to avoid having multiple copies of the file descriptor
+     */
+    FileDescriptorImpl &operator=(const FileDescriptorImpl &) = delete;
+
+    /**
+     * A move constructor allows the resources owned by an rvalue object to be moved into an
+     * lvalue without creating its copy.
+     */
+    FileDescriptorImpl(FileDescriptorImpl &&obj)
+    {
+        this->m_fd = obj.m_fd;
+        obj.m_fd   = FileDescriptor::invalid_descriptor;
+    }
+
+    /**
+     * Close socket on destructor (if it is still open)
+     */
+    ~FileDescriptorImpl() override { close(); }
+
+    /**
+     * @brief Gets file descriptor.
+     *
+     * File descriptor can be used:
+     * - in select(), poll() or epoll() to wait for events on this descriptor.
+     * - in send() family of functions to send data through the socket connection.
+     * - in recv() family of functions to receive data from the socket connection.
+     *
+     * @return File descriptor value.
+     */
+    int fd() override { return m_fd; }
+
+    /**
+     * @brief Closes file descriptor.
+     *
+     * If valid, closes file descriptor and then invalidates it.
+     *
+     * @return True on success and false otherwise (for example, if it was already closed)
+     */
+    bool close() override
+    {
+        printf("close(%d)\n", m_fd);
+        if (FileDescriptor::invalid_descriptor == m_fd) {
+            return false;
+        }
+
+        int rc = ::close(m_fd);
+        m_fd   = FileDescriptor::invalid_descriptor;
+
+        return (0 == rc);
+    }
+
+private:
+    int m_fd;
+};
+
+/**
+ * Wrapper class around sockadr (structure describing a generic socket address)
+ */
+class SockAddr {
+public:
+    virtual ~SockAddr()                             = default;
+    virtual const struct sockaddr *sockaddr() const = 0;
+    virtual const socklen_t &length() const         = 0;
+
+    struct sockaddr *sockaddr()
+    {
+        return const_cast<struct sockaddr *>(const_cast<const SockAddr *>(this)->sockaddr());
+    }
+    socklen_t &length()
+    {
+        return const_cast<socklen_t &>(const_cast<const SockAddr *>(this)->length());
+    }
+};
+
+/**
+ * Wrapper class around msghdr (structure describing messages sent by `sendmsg` and received by
+ * `recvmsg`)
+ */
+class MsgHdr {
+public:
+    virtual ~MsgHdr()                           = default;
+    virtual const struct msghdr *msghdr() const = 0;
+
+    struct msghdr *msghdr()
+    {
+        return const_cast<struct msghdr *>(const_cast<const MsgHdr *>(this)->msghdr());
+    }
+};
+
+/**
+ * Any object implementing the Socket interface is a FileDescriptor too.
+ * This way we give meaning to the action of "closing the socket".
+ */
+class Socket : public FileDescriptor {

```

Comment:
This relationship models **IS A**. I don't think that `socket` _is a_ `file descriptor`.
Socket might _have_ a file descriptor:
```cpp
class Socket{
FileDescriptor m_fd // in case FileDescriptor is a concrete class
FileDescriptor &m_fd // in case FileDescriptor is an interface
};
```
_Have a_ is weaker then _Is a_ and therefore preferable.

Reply:
As observed elsewhere, it it probably not useful to have FileDescriptor as a separate class, just move its functionality into Socket.

Reply:
This can be modeled in different ways. We could also think in a `Descriptable` interface if you feel like its semantics are more appropriate or closer to reality. This way `Socket` would extend `Descriptable` and its implementations would include a `FileDescriptorImpl` member. But we cannot move `FileDescriptor` into `Socket` because we'll need a `Timer` class (and maybe others in the future) that is also `Descriptable`

Reply:
Yes, Timer proves that FileDescriptor is needed.

Responding to the original remark: this ties in a bit with the discussion about handlers as class-to-be-derived-from or struct-to-be-assigned to. With an is-a relationship, you can allow both Socket and Timer to be passed into EventLoop without requiring special-casing each of them. The problem is that callbacks then get a FileDescriptor object and they can't do a lot with that, because they need a Socket or Timer or whatever to actually use it. With is-a relationship, you can at least do a dynamic cast to the correct type.

But even better is if the Socket/Timer object is part of the handler object, because then you immediately have the object you need. And then you don't need Socket to derive from FileDescriptor; instead, you can have a get_fd() function on the handler that retrieves the file descriptor from the socket's m_fd.

And at that point, maybe indeed, EventLoop doesn't need to see the FileDescriptor any more. EventHandler::get_fd() can just return an int.

So, I would go in this direction:
```c++
class EventHandler {
public:
    int get_fd() const = 0;
    bool handle_read_ready(EventLoop& loop) = 0;
    ...
};
```

And also, that the Connection class is a handler (or there's a parallel class for it), where the handler reads in all available bytes and calls a secondary configured handler with a Buffer as argument.